### ISSUE TITLE
Exporting to SPSS, limited string width

### ISSUE THREAD

**Issue by markboukes** [amcat/amcat#500]
Hi there,

I have some problem with exporting my articles from AMCaT to an SPSS file. For the string variables that are being exported, e.g., "text" from LexisNexis-files, it automatically shortens the main text to a maximum of 765 characters. This of course is not enough to catch the whole text.

Is there anything to do about this?

Best,
Mark.

![spssamcat](https://cloud.githubusercontent.com/assets/12925601/14637170/dc688874-062f-11e6-86ea-19107940cdf4.jpg)

**Comment by martijnbastiaan** [comment]
Hi!

Could you attach the SPSS file AmCAT generated?

**Comment by markboukes** [comment]
Yes, would like to, but this webpage does not allow to attach it, also not in a .zip file. Can i e-mail it to you?

**Comment by martijnbastiaan** [comment]
Sure! martijn [at] hmbastiaan [dot] nl.

**Comment by martijnbastiaan** [comment]
PSPP does not support long lines when using the ["DATA LIST LIST"](https://www.gnu.org/software/pspp/manual/html_node/DATA-LIST-LIST.html#DATA-LIST-LIST) command. Our best bet is to generate an ODS file and use [GET DATA](https://www.gnu.org/software/pspp/manual/html_node/GET-DATA-_002fTYPE_003dGNM_002fODS.html#GET-DATA-_002fTYPE_003dGNM_002fODS).

**Comment by vanatteveldt** [comment]
What happens if you export the CSV and read that in using the text import wizard in SPSS?

**Comment by markboukes** [comment]
That does not work either. Tried it already, but gives very strange output in SPSS.

**Comment by martijnbastiaan** [comment]
I suspect PSPP uses [DATA LIST FREE](https://www.gnu.org/software/pspp/manual/html_node/DATA-LIST-FREE.html#DATA-LIST-FREE) internally, which interprets newlines as new rows..

**Comment by martijnbastiaan** [comment]
I've installed and tested SPSS itself, but it seems to have very poor support (if at all?) for newlines. Imported Excel sheets containing newlines produce weird characters, while a CSV with a newline completely breaks import support. I suppose the best way to go about this is to replace all newlines with spaces. Probably like:

```python
def _get_lines(text):
    for line in (line.strip() for line in text.split("\n")):
        if line:
            if line.endswith("."):
                yield line
            else:
                yield "." + line

def get_spss_text(text):
    return " ".join(get_lines(text))

get_spss_text(article.text)
```

This allows humans to read it more easily (separation between header and body in articles) and should still allow for analysis. I don't think perfect information transfer is needed for SPSS?

@vanatteveldt ?

**Comment by vanatteveldt** [comment]
I'm inclined to drop it here. We export good csv, and if SPSS can't import it properly it's not our problem.

I've also talked to the user needing this today, and it seems that she can use AmCAT directly for the text analysis she needs. @markboukes is that correct?

@martijnbastiaan , if you want to, feel free to implement, otherwise I would just close the issue.

**Comment by martijnbastiaan** [comment]
It's probably simple enough to implement a replacement function, and SPSS "support" will probably make AmCAT more attractive. I'll add a warning as well.

**Comment by markboukes** [comment]
Great, @martijnbastiaan! It would save me a lot of pain, and back and forths between different programs. Besides that, it will help students amazingly to just have it immediately in SPSS.

Besides the function the user @vanatteveldt talks about above, i can imagine different functions for which other solutions are not so easy to find. One that i am personnally confronted with at the moment :)

So, i really hope you can make this happen!!

Thanks in any way.

**Comment by martijnbastiaan** [comment]
https://github.com/amcat/amcat/commit/1701dee1459f7a8c30e9b4bb2c31d664c5d34f71 added support for:

* Really long lines (32767 characters per cell)
* Files containing newlines (by replacing them with single dots)

This has been pushed to the development version of AmCAT. It would be good to also backport it to AmCAT 3.4, although that would take some more work which I'll probably do over the weekend.

**Comment by martijnbastiaan** [comment]
Current limitations are (and these won't be changed, most probably):

* 32767 characters (or bytes?) per cell
* Tabs in texts are removed and replaced by spaces
* Newlines in texts are removed and replaced by dots

The UI will display the limitations if a user tries to export to SPSS. Other limitations may exist, but I'm not aware of any.

**Comment by martijnbastiaan** [comment]
Warning added in https://github.com/amcat/amcat/commit/8370b9c1cb10c2f5e952c108301b058c555bd289.

![screenshot from 2016-04-23 00-53-27](https://cloud.githubusercontent.com/assets/5658854/14755877/35467b72-08dd-11e6-840c-e5daafb5fd0c.png)

**Comment by martijnbastiaan** [comment]
The SPSS exporter has been backported to 3.4, and passes the automated tests. In production a strange thing occurs: pspp gets launched, but the process locks up and amcat keeps waiting for an answer. Internally, we use `communicate()` so it's probably not a stdout/stdin deadlock.

**Comment by markboukes** [comment]
Hi guys, when i'm trying to export anything to SPSS now, i get the following error:

{
  "status": 500, 
  "message": "500 : An Error has occurred", 
  "description": "The server encountered an internal error or misconfiguration and\n                     was unable to complete your request.", 
  "details": "KeyError: <type 'unicode'>", 
  "error": true
}

This happened already last week and todayt, with a big (n=15,000) as well as a smaller dataset (n=900), and both in Firefox and Chrome.

**Comment by martijnbastiaan** [comment]
Fixed in release-3.4 (https://github.com/amcat/amcat/commit/786daa7b19b082776a6d8ee01bb83bb744b29e9f). Happy SPSS exporting :).
